### ISSUE TITLE
P3 code design restructuring

### ISSUE THREAD

**Issue by bartgol** [E3SM-Project/scream#478]
I've been discussing this on the slack channel, but a) I don't think long conversations belong there, and b) conversations get easily lost in the channel history.

Following the realization that one file was taking 4m30s to compile on GPU, I've been taking a closer look to P3 code organization. After giving it some thought, I want to propose a different organization of the code. However, I wasn't (and still am not) involved with its development, and I don't really like to step in someone else's turf. So I'd like to hear the opinion of p3 main devs: @jgfouca , @jeff-cohere , @PeterCaldwell , @pressel , just to ping a few. I'm leaving Andrew out cause I know he's not much interested in this kind of conversations.

Here are my thoughts:

1) The types in `Functions` should be carved out and put in a separate file (something like `p3_types.hpp`). They do not have to be in a class or namespace either, imho. The scope `scream::p3` is already plenty to avoid name clashes.

2) The `Functions` class should probably be just a namespace. There's nothing in Functions that *needs* a class pattern: no inheritance, no extension. If the goal is to encapsulate in a scope, then namespace is the proper C++ way. Namespaces have one big advantage compared to classes: if you need to extend your list of functions, you can do it _anywhere_, while with classes you need to modify the whole class interface, even if 99% of the customers of that class do not care about your new function.

3) We should probably put all the `p3_xyz.cpp` files in a separate folder, like `src/physics/p3/eti`. This will keep the number of items in the p3 folder limited, which helps visually when listing the content of the folder. Maybe we could also relegate the impl files to a separate folder like `src/physics/p3/impl/`. There are 35 `p3_xyz_impl.hpp p3_xyz.cpp` pairs in `src/physics/p3`, so adding these two subfolders might help navigate the folder. Notice that the location of the impl and cpp files is irrelevant for all customers of those functions (only cmake and `p3_functions.hpp` itself care about it).

4) The file I mentioned at the beginning, that takes a long time to compile, is `p3_functions_f90.cpp`. Among others, it contains F90->C bridges (C callable from F90). Since we no longer call C from F90, on slack I was floating the idea of adding some ifdef logic to compile these bridges if explicitly requested (might be useful if something breaks, for debugging purposes). However, Jim made me notice that they are also used in some BFB unit tests, as a way to pack the test input data from host to device. A suggestion that came up was to also break p3_functions_f90.*pp into several files, each including only the C++ p3 function that it needs (somewhat mirroring the splitting of C++ files). I think I like this idea. We could relegate these bridges to a `f2c` or `f2c_bridges` folder.

Ok, I am starting to lose momentum for writing, so I'll stop here and see if you guys have any comment.

**Comment by jeff-cohere** [comment]
Hey @bartgol . I think it's a great idea to have a conversation about this stuff. I'm somewhat consumed with the p3 struct reorg at the moment, but after I get that in, I'll take a closer look at this.

**Comment by jgfouca** [comment]
@bartgol , good points. I agree with all of them except maybe (2) and (1) to a lesser extent... I'm a bit unsure of all the consequences of changing Functions to a namespace. My hunch is that the template/type situation will not be as convenient as it is now.

**Comment by jgfouca** [comment]
Also, I've assigned myself just to make sure this doesn't get lost.

**Comment by jeff-cohere** [comment]
Jim raises a good point about (2). It would be interesting to see how our code might look if we used a namespace rather than a struct (which allows us to associate types with it). The newfangled P3 structs can serve as examples for both approaches once PR #465 gets merged.

I'm all for breaking up translation units if it helps build times. Maybe we could take a piecemeal approach and start with Luca's `eti/` subdirectory idea, and see if that doesn't solve the visual navigation problem by itself. I also like the idea of an `f2c` folder for the Fortran bridge.

**Comment by PeterCaldwell** [comment]
I like the idea of putting cpp and impl files in their own subdirectories. Breaking p3_functions_f90.cpp into a bunch of files and putting them in their own subdirectory in particular sounds like a good idea. It may in particular help with writing a code that auto-generates boilerplate for us. I'm out of my depth on ideas 1 and 2.

One thing to weigh against for these changes is that time spent implementing them is time spent not doing other things. Are these changes worth it? I suspect so, but we should keep perspective. The other thing to keep in mind is what we should do for SHOC so we get the right structures from the get go.

**Comment by bartgol** [comment]
@PeterCaldwell , the time-resource concern is always a pertinent one. I don't know if it's worth it. I guess that depends on how much people will develop in P3, or at least navigate the code. I believe a better organization makes code navigation faster, but how much I can't quantify (especially I can't quantify for people other than me).

@jgfouca Regarding the templates, by turning each static member function into a free function, we would have to template each funciton on scalar/device. So let's see how it looks. Method like `cloud_water_conservation` or `calc_first_order_upwind_step` will turn into

```
// No need to template this on device type
template<typename Scalar>
KOKKOS_FUNCTION
void cloud_water_conservation (
   const Spack<Scalar>& qc, const Scalar dt, 
   Spack<Scalar>& qcaut, Spack<Scalar>& qcacc,
   Spack<Scalar>& qccol, Spack<Scalar>& qcheti,
   Spack<Scalar>& qcshd, Spack<Scalar>& qiberg,
   Spack<Scalar>& qisub, Spack<Scalar>& qidep,
   const Smask& context = Smask(true) );

template<typename Scalar, typename Device>
KOKKOS_FUNCTION
void calc_first_order_upwind_step(
      const uview_1d<const Spack<Scalar>,Device>& rho,
      const uview_1d<const Spack<Scalar>,Device>& inv_rho, // 1/rho
      const uview_1d<const Spack<Scalar>,Device>& inv_dzq,
      const MemberType<Device>& team,
      const Int& nk, const Int& k_bot, const Int& k_top, const Int& kdir, const Scalar& dt_sub,
      const uview_1d<Spack<Scalar>,Device>& flux,
      const uview_1d<const Spack<Scalar>,Device>& V,
      const uview_1d<Spack<Scalar>,Device>& r);


```
Where in `p3_types.hpp` you would have

```
template<typename Scalar>
using Spack = scream::pack::SmallPack<Scalar>;

// Note: <int> is just a place holder. All we want is the size of the small pack, who cares about the stored type.
using SmallMask = scream::pack::Mask<Spack<int>::n>;

template<typename Device>
using MemberType = KokkosTypes<Device>::MemberType;

template<typename Scalar, typename Device>
using view_1d = KokkosType<Device>::template view_1d<Scalar>;

template<typename Scalar, typename Device>
using uview_1d = typename ko::template Unmanaged<view_1d<Scalar,Device>>;
```

So yeah, the individual functions signatures would be a bit longer (not much if one is ok with `typename S, typename D`). But would allow to break up everything in single files. Right now, `p3_functions_f90.*pp` _does_ need p3_functions.hpp (all of it), which on GPU includes _all_ the functions impls.

Side note: why are we templating the whole class on a scalar? Our 'Real' type is already a configurable option. Unless we plan on running some functions with one scalar type and some other with another scalar type...

**Comment by jgfouca** [comment]
@bartgol , what about ETI? Wouldn't we have to instantiate every function instead of just the struct?

**Comment by bartgol** [comment]
Well, that's the same thing that happens now. We are instantiating the Functions. E.g, in `p3_ice_collection.cpp`, Instead of having

`template struct Functions<Real, DefailtDevice>;`

you will have

`template p3_ice_collection<Real,DefaultDevice>(...);`

Not a big change, I think.

**Comment by jeff-cohere** [comment]
I think Jim is referring to the fact that there will be more such statements--one per function instead of the whole shebang. But this is necessary if we actually want to parallelize the instantiations, I think.

**Comment by bartgol** [comment]
Well, we already have one ETI statement in each cpp file. It is the same statement, yes, but does that really make the code (c)leaner? I actually had to think for a moment at what the compiler is doing when you ETI a struct only giving impl for one function. Of course, it does what you want, but I would argue that, if your goal is to ETI (being able to promote a noun to a verb is my fav English feature) a single function, then it is much clearer if you are indeed ETI-ing a single function, rather than the struct (especially for devs who are not super familiar with template programming/compilation mechanics).

**Comment by jgfouca** [comment]
@bartgol you're right, ETI doesn't really make much difference either way. My main concern now is that the function signatures do not look as nice without being able to use the precomputed typedefs in the struct.

**Comment by bartgol** [comment]
Another thing we could do (as a piece of a/the solution) would be to compile `p3_functions_f90.*pp` _only if_ tests are enabled. In a CIME build, where we only want to build e3sm.exe, there's no point in building it (since the F->C bridges are not used, and the tests are not built).

**Comment by PeterCaldwell** [comment]
@jgfouca and @bartgol - I thought we addressed this, but p3_functions_f90.cpp is still 3500 lines long so maybe not? Should we make this a wishlist item?

**Comment by bartgol** [comment]
We never touched the f90 bridging. The file you mentioned would really benefit from being broken down (like the C++ functions impl), so that it can be compiled in parallel, which would really help GPU build times.

**Comment by bartgol** [comment]
Regarding the organization of files and the struct-vs-namespace debate, I just re-read the conversation, and still think that a namespace would be more appropriate and easier to maintain/read. I'm going to assign low priority.

**Comment by AaronDonahue** [comment]
@bartgol , after our massive restructuring of the P3 directories and files has this issue from 1.5 yrs ago been addressed?  Can we close it?
